### PR TITLE
DRAFT: Basic focusgroup behavior

### PR DESCRIPTION
**Pull request by robarbms**
# Pull Request

## 📖 Description

Looking at adding keyboard arrowing through a behavior.

## 👩‍💻 Reviewer Notes

I've added focusgroup behavior to non-interactive calendars for demo purposes. This would not ship.

## 📑 Test Plan

Need to build out unit tests once I have a working model.

### General

- [ ] I have included a change request file using `$ yarn change`
- [ ] I have added tests for my changes.
- [x] I have tested my changes.
- [ ] I have updated the project documentation to reflect my changes.
- [ ] I have read the [CONTRIBUTING](https://github.com/Microsoft/fast/blob/master/CONTRIBUTING.md) documentation and followed the [standards](https://www.fast.design/docs/community/code-of-conduct/#our-standards) for this project.

### Component-specific

- [ ] I have added a new component
- [ ] I have modified an existing component
- [ ] I have updated the [definition file](https://github.com/Microsoft/fast/blob/master/packages/web-components/fast-components/CONTRIBUTING.md#definition)
- [ ] I have updated the [configuration file](https://github.com/Microsoft/fast/blob/master/packages/web-components/fast-components/CONTRIBUTING.md#configuration)

## ⏭ Next Steps

Finalize options and create a standard api.
Improve logic when there are no items with overlap.
Work through the focusgroup spec and better align with it.
Unit tests.
Polyfill behavior, use the browser functionality when available, fallback to behavior when not available.

### CODE REVIEW

**Comment by chrisdholt** [comment]
Adding a link to the relevant explainer for `focusgroup` which is being considered as a new attribute: https://github.com/MicrosoftEdge/MSEdgeExplainers/blob/main/Focusgroup/explainer.md

Ideally the above is only used if the native browser behavior is not supported. In the case `focusgroup` is supported, we would fall back to that. @robarbms feel free to correct 👍 

@travisleithead as FYI

**Comment by robarbms** [comment]
That is correct. It would act as a polyfill. As I understand it, there isn't currently a browser that supports the focusgroup yet so I'm not checking for it.

https://caniuse.com/?search=focusgroup

I will need to use the browser functionality instead, basically just outputting `focusgroup` into the html template, once it's available.

File: packages/web-components/fast-element/src/templating/focusgroup.ts [hunk 1]
```diff
@@ -0,0 +1,428 @@
...
+        index: number;
+    };
+
+    /**
+     * Gets the elements under the focusgroup that are focusable
+     * @param node - root node to find focusable items under
+     * @returns - an array of html elements that are focusable to arrow between
+     * @public
+     */
+    public getFocusItems(node: HTMLElement = this.target): HTMLElement[] {
```
Comment (scomea) [inline packages/web-components/fast-element/src/templating/focusgroup.ts:107]: Should we be using tabbable here like we do in dialog?

File: packages/web-components/fast-element/src/templating/focusgroup.ts [hunk 2]
```diff
@@ -0,0 +1,428 @@
...
+    }
+
+    /**
+     * keydown handler for the parent element
+     * @param event - keyboard event
+     * @returns - should bubble
+     * @public
+     */
+    public handleKeydown(event: KeyboardEvent): boolean {
+        const { target } = event;
```
Comment (scomea) [inline packages/web-components/fast-element/src/templating/focusgroup.ts:368]: Check for defaultPrevented before acting, and can implementors easily get first crack at handling the keystroke first if they wish?

File: packages/web-components/fast-element/src/templating/focusgroup.ts [hunk 3]
```diff
@@ -0,0 +1,428 @@
...
+            );
+    }
+
+    /**
+     * keydown handler for the parent element
+     * @param event - keyboard event
+     * @returns - should bubble
+     * @public
+     */
+    public handleKeydown(event: KeyboardEvent): boolean {
```
Comment (scomea) [inline packages/web-components/fast-element/src/templating/focusgroup.ts:367]: How does returning a bool here work?  The event listener is hooked up via addEventListener and not a template binding and the bool only works in the latter case.  Shouldn't we always mark the event with preventDefault() when we move focus as a result?

File: packages/web-components/fast-element/src/templating/focusgroup.ts (hunk 1 above)
Comment (robarbms) [reply packages/web-components/fast-element/src/templating/focusgroup.ts:107]: Thanks! I was using a hard coded list that I wasn't happy with and did not want to maintain. I've copied your code from dialog for identifying tabbable elements.

**Pull request closed without merging**